deleteat!: Handle unassigned array elements

Keno opened the pull request:
This function was erroring on unassigned array elements. This is
the minimal fix to resolve that issue. I tried something more
generic, but it turns out all of our Array code basically assumes
that all elements are assigned. Even basic things like `==`, `hash`,
etc. error for arrays with unassigned elements. I think we may want
to strongly consider removing the ability to unassign elements in
arrays in favor of Union{Nothing, Some{T}} arrays, which our existing code
handles fine and should be able to be made to have equivalent performance.

Seelengrab commented:
> I think we may want to strongly consider removing the ability to unassign elements in
> arrays in favor of Union{Nothing, Some{T}} arrays, which our existing code
> handles fine and should be able to be made to have equivalent performance.

I think this in part boils down to whether we allow `undef` array initializer or not, e.g. see discussion in #42620, #9147 and #130.

vtjnash commented on base/array.jl:
```diff
@@ -216,11 +216,15 @@ length(a::Array) = arraylen(a)
 
-function isassigned(a::Array, i::Int...)
+function isassigned(a::Array{T}, i::Int...) where {T}
     @inline
     ii = (_sub2ind(size(a), i...) % UInt) - 1
     @boundscheck ii < length(a) % UInt || return false
-    ccall(:jl_array_isassigned, Cint, (Any, UInt), a, ii) == 1
+    if allocatedinline(T) && (T isa DataType) && datatype_pointerfree(T)
+        return true
+    else
+        return ccall(:jl_array_isassigned, Cint, (Any, UInt), a, ii) == 1
+    end
```
This should not need to be changed (it is already handled in the optimizer)

vtjnash commented:
> favor of Union{Nothing, Some{T}} arrays, which our existing code
handles fine and should be able to be made to have equivalent performance

We do not currently have the code to handle that fine. We probably should, but it would mean making this array unable to contain `#undef`, which we currently permit.

Keno committed:
deleteat!: Handle unassigned array elements

This function was erroring on unassigned array elements. This is
the minimal fix to resolve that issue. I tried something more
generic, but it turns out all of our Array code basically assumes
that all elements are assigned. Even basic things like `==`, `hash`,
etc. error for arrays with unassigned elements. I think we may want
to strongly consider removing the ability to unassign elements in
arrays in favor of Union{Nothing, T} arrays, which our existing code
handles fine and should be able to be made to have equivalent performance.
